### DISCUSSION TITLE
Why not split this into many small plugins?

### DISCUSSION THREAD

**Discussion by RobIsHere** [vikejs/vike#979] [Discussion]
### Description

I'm getting lost in the magic and obfuscation that is done by this package. Config changes magically. Server entry file doesn't exist but is loaded by some loading-magic plugin that can also be manually configured. And so on and on. I've read a lot of old discussion and issues to find links to detached docs (not linked from official docs) randomly that help solve your problem.
You just keep hitting walls...

At the moment I'm pretty sure It would have saved me half of the time if I had just started with the vite low level ssr api.

There are also a lot of great things in this library - why don't you put the build stuff into a dedicated plugin? So it gets exchangable or you even do not have to use it at all.
(So I could just put my server.ts entry file into my code, import stuff from there and build a self contained server build like we did it for a decade without need for copying gigabytes into a docker image)

Another priority is to factor out this magic server loader plugin that confuses everyone - which someone can install only if needed. Having a real server.ts entry file would solve a lot of problems for newcommers.

Nice to have also : the other things IMHO should be split in two plugins - one without router and one with only the routing. Maybe even a third one with server-routing capabilities. The thing is I need to initialize redux on the server and therefore I need redux-integrated routing. Which I could just plug in. I know I can do this now by just ignoring existing routing code, so thats not so much a priority.

So ideally you end up with 4-5 small plugins that work together. And if the build doesn't fit your needs, you just uninstall that and write your own code.

The current situation is that I have only the choice to put even more time into it to get it running somehow maybe with vite and vercels ncc (just moving away from vercel) like suggested by someone somewhere. And the other option is to start from zero from vite or vavite (that's what I decided).

With plugins I could keep the good things.

**Comment by brillout** [comment]
I agree that the DX should be a lot smoother around this.

Before we talk about solutions, let's discuss the issues.

Are you using Yarn? If you don't then it should just work. If you aren't using Yarn, then let me know what the issues are and I'll dig into them.

As for self-contained server builds, what's your deployment strategy?

> Config changes magically.

I'm not sure what you mean here.

> And so on and on

Please elaborate.

Moving to GitHub Discussions as it isn't a bug per se, but I do take these issues seriously.

Let's continue the discussion and tackle the issues one by one.

**Reply by brillout** [reply]
@RobIsHere I believe the [v1 release](https://github.com/brillout/vite-plugin-ssr/issues/985) will address all the issues you encountered. But let me know if you believe something is missing. (I've also improved the readme of https://github.com/brillout/vite-plugin-import-build. As for making `vite-plugin-import-build` optional: I share the sentiment but I think it's possible to address your issues without structural changes.)

**Reply by RobIsHere** [reply]
I'm not using yarn pnp but yarn classic. 

I realized some additional things: dev and production must use the same tooling. That means having esbuild and rollup is a problem. Another problem is having esbuild by hmr in dev and not having it in production on locally linked node_modules which causes all the deployment issues.

One thing that someone didn't understand at vite is that the fact that nodejs now supports modules does not mean that anyone is supposed to copy thousands of source files onto servers that get requested and ts-transpiled there one by one.

No other sane programming language would do anything like that. All the windows DLL and linux shared libraries are linked together from lot of files before the libraries (js-slang: bundles) are deployed. It's equivalently just normal to link the js/ts files before deploying them. No one would directly use the object files (that's the file-by-file output of the c/c++ compiler), that's just unthinkable. But exactly this strange procedure promotes vite for production. Don't know how anyone could expect this to work well, there are too many beginners in JS.

In fact I reconfigure the build to bundle the dependencies with esbuild in production. That's also the point where you quesition yourself if vite is still an advantage at all or just bloat. It looks like that was different until version 2, maybe that explains the popularity.

The issues with the ssr plugin are somehow interdependent on the things above. Sorry for taking your time, but I needed some time to realize the dimension of the problems with vite and that it's vite that's the real problem and not so much the plugin.

**Reply by brillout** [reply]
Fundamentally, Vite does "on-demand transpiling", which means:
 - During dev, if you go to `http://localhost:3000/some-page` then only the code needed for that page is transpiled. If you have 100 pages then only a tiny fraction of your app is transpiled to render that page. This makes developing *much* faster.
 - During build, your whole app needs to be transpiled. Inherently, this means the transpilation process is quite different between dev and build/production.

It isn't possible to have it both and I'd argue it's better to have a *much* faster dev DX than the guarantee that dev and build will lead to the exact same code. I agree that the dev/build discrepancy makes things less predictable, but I think it's very well worth it. The fast dev DX of modern JavaScript (especially with Vite) is actually something I like about the current JavaScript status quo.

I agree a lot of things are rough in JavaScript land (which is also *very* frustrating to me and a reason why I've created vite-plugin-ssr), but I think the dev/build discrepancy is a minor issue and that the problems are elsewhere.

> Sorry for taking your time

I actually appreciate that you're sharing your thoughts and frustation. Hopefully (and I believe so) most issues will be adressed once the v1 is released: https://github.com/brillout/vite-plugin-ssr/labels/v1%20release%20%3Astar2%3A and most notably https://github.com/brillout/vite-plugin-ssr/issues/562, https://github.com/brillout/vite-plugin-ssr/issues/984, and https://github.com/brillout/vite-plugin-ssr/issues/983.

> I reconfigure the build to bundle the dependencies with esbuild in production.

That's actually currently the best way to do it. I expect this works for you, or are you still having issues?

Let me know if you're having any other issue.

**Reply by RobIsHere** [reply]
Honestly: who has really experienced speed issues with "slow" build tools? Everyone is acting like he had a google-size repo at home that that monster-repo needs nothing less but the fastest devops. Developers becomming insane because of half-baked solutions is just collateral damage.

In reality almost no one has a monster-repo and if someone has a monster-repo he did so many things wrong, that I wouldn't consider him able to judge things like these.

You can prebundle non-local node_modules packages which should improve the speed to a level that simply makes it not relevant if the remaining 100kb of sourcecode is compiled by a tool that is 10  or 20 times slower.

If I can have a 'just works' experience I would always prefer a preconfigured webpack or rollup that just works since years over an swc or esbuild with the 1000 quirks in it.

With the new "tools" in 0.x state I really couldn't tell my customer in the morning if I get ready midday. I just didn't know if the next.js compiler used in it will work. That's because you need an experimental flag to make it work in their own monorepo product - which they tell you after luring you into it. That's the current state of insanity we reached. You could become religious again while watching the build tools build or not build. Because you know if it doesn't work, I have a huge problem. I don't remember anything like this with c/c++ or java compilers. Not even with php interpreters.

If you do want to create something better out off vite-plugin-ssr make a well maintained working solution with proven tooling please! 

From my perspective it's the point in time with a bigger chance for you to establish a better thing. I'm very disappointed with next.js - which I tried after vite-plugin-ssr - because of a lot of things (caching by default without respecting danger of loosing data by continuing work on stale data e.g, bad working monorepo tooling, big ball of mud sourcecode style in their sources).

A lot of the issues you linked are pointing in the right direction. Good luck!

**Reply by brillout** [reply]
Build time isn't the issue, it's dev speed that matters. And it does quickly matter.

I agree robustness is something that is lacking today, and I do intend to make vite-plugin-ssr a robust tool. It will take a little while though.

(FYI I've a design in mind to remove `vite-plugin-import-build` which will solve some of the problems you encountered.)
